handle syntax errors more gracefully

aaronjanse opened the pull request:
Fixes https://github.com/nix-community/rnix-parser/issues/33.

I don't know enough about `rowan` to cleanly improve error handling, but I thought I'd at least share the patch I've been using locally while implementing rnix-lsp autocomplete.

Feel free to re-use parts of this patch for a different PR, push directly to this PR, or something else.

Ma27 commented on src/parser.rs:
```diff
@@ -291,7 +301,15 @@ where
+                let end = self.finish_error_node();
+                self.errors.push(ParseError::Expected(
+                    TextRange::new(start, end),
+                    "attribute".to_string(),
```
I guess we should reflect here that strings etc. are also OK.

aaronjanse replied:
What should be the exact message used? Maybe `attribute or string`? Although I guess a string does represent an attribute here? Please correct me if I'm wrong

Ma27 reviewed (commented):
LGTM mostly, nice work! :) 

Would you mind resolving the last comment and rebase?

aaronjanse commented:
I pushed a somewhat significant change to behavior. The change matters only for attr path now. I also added an "expected semicolon" error node and updated tests

oberblastmeister commented:
You need to run this with a fuzzer. Adding better error handling can easily screw up the parser, and I got an out of memory error relatively quickly. I would prefer more general error handling (removing `expect_peek_any`). It would be also be good to see the effectiveness of the error handling by hooking it up with `rnix-lsp`. Maybe a mono repo would be good for this?

aaronjanse commented:
Wow, good catch @oberblastmeister. The fuzzer is amazing, and after seeings its result, I agree that we need a more general  solution here. I'll start working on that. Thank you!!

oberblastmeister commented:
@aaronjanse I already did a lot of work here https://github.com/nix-community/rnix-parser/pull/35. Also, I really think we should work on hooking up the new version of `rnix` to `rnix-lsp` before we do improved error handling. This way we can try out the error handling in real time.

aaronjanse commented:
> @aaronjanse I already did a lot of work here https://github.com/nix-community/rnix-parser/pull/35

Oh, sorry about that! I'm still catching up after being away for a while. If you submit a PR for the error handling that was in #35, I'll close #120 in favor of your PR.

> I really think we should work on hooking up the new version of rnix to rnix-lsp before we do improved error handling. This way we can try out the error handling in real time.

As in, you think we should update rnix-lsp to use the latest nix-parser version? I'd be willing to work on doing that

oberblastmeister commented:
> As in, you think we should update rnix-lsp to use the latest nix-parser version? I'd be willing to work on doing that

Yes! I think that would be great.

aaronjanse commented:
Sounds good, will do!
